### PR TITLE
[Improvement] Better unit test.

### PR DESCRIPTION
**Pull request by grimoire**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

The unit tests might not release resources if the test failed. Create fixtures for the resources.

## Modification

- Fix rewriter staticmethod
- Better backendswitchwrapper
- resources fixture.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: csrc/mmdeploy/backend_ops/tensorrt/roi_align/trt_roi_align.cpp [hunk 1]
```diff
@@ -202,7 +202,7 @@ nvinfer1::IPluginV2 *TRTRoIAlignCreator::createPlugin(
     if (field_name.compare("mode") == 0) {
       int data_size = fc->fields[i].length;
       const char *data_start = static_cast<const char *>(fc->fields[i].data);
-      std::string poolModeStr(data_start, data_size);
+      std::string poolModeStr(data_start);
```
Comment (lvhan028) [inline csrc/mmdeploy/backend_ops/tensorrt/roi_align/trt_roi_align.cpp:205]: may use snake case

File: mmdeploy/core/rewriters/function_rewriter.py [hunk 2]
```diff
@@ -72,7 +73,16 @@ def _set_func(origin_func_path: str,
...
+
+    is_static_method = False
+    if method_class:
+        origin_type = inspect.getattr_static(module_or_class, split_path[-1])
+        is_static_method = isinstance(origin_type, staticmethod)
+
+    if is_static_method:
+        exec(f'{origin_func_path} = staticmethod(rewrite_func)')
+    else:
+        exec(f'{origin_func_path} = rewrite_func')
```
Comment (AllentDan) [inline mmdeploy/core/rewriters/function_rewriter.py:85]: why do we make difference between static methods and other methods? Removing `self` for beginners might be confusing compared with original way of rewriting.

Comment (grimoire) [reply mmdeploy/core/rewriters/function_rewriter.py:85]: 1. static method does not have self 
https://github.com/open-mmlab/mmdetection3d/blob/47285b3f1e9dba358e98fcd12e523cfd0769c876/mmdet3d/core/bbox/coders/fcos3d_bbox_coder.py#L103
2. static method need special treatment, or it won't recover after context exit.

Comment (AllentDan) [reply mmdeploy/core/rewriters/function_rewriter.py:85]: I see.

File: tests/test_codebase/test_mmcls/test_classification.py [hunk 3]
```diff
@@ -73,58 +81,57 @@ def test_init_pytorch_model(from_mmrazor: Any):
...
-def test_visualize(backend_model):
-    input_dict, _ = task_processor.create_input(img, input_shape=img_shape)
+def test_visualize(task_processor, backend_model, tmp_path, img, model_inputs):
+    input_dict, _ = model_inputs
     results = task_processor.run_inference(backend_model, input_dict)
-    with TemporaryDirectory() as dir:
-        filename = dir + 'tmp.jpg'
-        task_processor.visualize(backend_model, img, results[0], filename, '')
-        assert os.path.exists(filename)
+    filename = str(tmp_path / 'tmp.jpg')
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmcls/test_classification.py:118]: is it OK to concat string in this way? Better use `os.path` to join directories.

File: tests/test_codebase/test_mmcls/test_classification_model.py [hunk 4]
```diff
@@ -1,123 +1,110 @@
...
         imgs = torch.rand(2, 3, IMAGE_SIZE, IMAGE_SIZE)
-        results = self.end2end_model.forward_test(imgs)
+        results = end2end_model.forward_test(imgs)
         assert isinstance(results[0], np.ndarray)
 
-    def test_show_result(self):
+    def test_show_result(self, end2end_model, tmp_path):
         input_img = np.zeros([IMAGE_SIZE, IMAGE_SIZE, 3])
-        img_path = NamedTemporaryFile(suffix='.jpg').name
+        img_path = str(tmp_path / 'tmp.jpg')
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmcls/test_classification_model.py:58]: here also

File: tests/test_codebase/test_mmdet3d/test_mmdet3d_models.py [hunk 5]
```diff
@@ -113,30 +117,34 @@ def test_pointpillars_scatter(backend_type: Backend):
...
 
 
-def get_centerpoint_head():
-    from mmdet3d.models import builder
-    model_cfg.centerpoint_model.pts_bbox_head.test_cfg = model_cfg.\
-        centerpoint_model.test_cfg
-    head = builder.build_head(model_cfg.centerpoint_model.pts_bbox_head)
-    head.requires_grad_(False)
-    return head
+# @pytest.fixture
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmdet3d/test_mmdet3d_models.py:130]: should we remove this ut?

File: tests/test_codebase/test_mmocr/test_mmocr_models.py [hunk 6]
```diff
@@ -506,7 +503,7 @@ def test_sar_model(backend: Backend, decoder_type):
     pytorch_model.cfg = sar_cfg
     patched_model = patch_model(
         pytorch_model, cfg=deploy_cfg, backend=backend.value)
-    onnx_file_path = tempfile.NamedTemporaryFile(suffix='.onnx').name
+    onnx_file_path = str(tmp_path / 'tmp.onnx')
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmocr/test_mmocr_models.py:506]: does this work for windows?

File: tests/test_codebase/test_mmcls/test_classification.py (hunk 3 above)
Comment (grimoire) [reply tests/test_codebase/test_mmcls/test_classification.py:118]: [tmp_path](https://docs.pytest.org/en/7.2.x/reference/reference.html#tmp-path) is a builtin fixture in pytest which return a [pathlib.Path](https://docs.python.org/3/library/pathlib.html#pathlib.Path) object.

File: tests/test_codebase/test_mmocr/test_mmocr_models.py (hunk 6 above)
Comment (grimoire) [reply tests/test_codebase/test_mmocr/test_mmocr_models.py:506]: [tmp_path](https://docs.pytest.org/en/7.2.x/reference/reference.html#tmp-path) is a builtin fixture in pytest which return a [pathlib.Path](https://docs.python.org/3/library/pathlib.html#pathlib.Path) object.

**Review by AllentDan — APPROVED** [review]

**Review by RunningLeon — APPROVED** [review]
LGTM

**Pull request merged**